src/main/java/com/flightstats/hub/dao/aws/s3verifier/MissingContentFinder.java

```diff
@@ -0,0 +1,121 @@
+                                @Named(ContentDao.SINGLE_LONG_TERM) ContentDao s3SingleContentDao,
+                                VerifierConfig verifierConfig,
+                                StatsdReporter statsdReporter,
+                                @Named(NamedDependencies.S3_VERIFIER_QUERY_THREAD_POOL) ExecutorService queryThreadPool) {
+        this.spokeWriteContentDao = spokeWriteContentDao;
+        this.s3SingleContentDao = s3SingleContentDao;
+        this.verifierConfig = verifierConfig;
+        this.queryThreadPool = queryThreadPool;
+        this.statsdReporter = statsdReporter;
+    }
+
+    public SortedSet<ContentKey> getMissing(MinutePath startPath, MinutePath endPath, String channelName) {
+        TimeQuery timeQuery = buildTimeQuery(channelName, startPath, endPath);
+        CompletableFuture<QueryResult> spokeQueryResults = getContentKeyFuture(() -> spokeWriteContentDao.queryByTime(timeQuery));
+        CompletableFuture<QueryResult> s3QueryResults = getContentKeyFuture(() -> s3SingleContentDao.queryByTime(timeQuery));
+
+        try {
+            CompletableFuture
+                    .allOf(spokeQueryResults, s3QueryResults)
+                    .get(calculateQueryTimeout(startPath, endPath), verifierConfig.getBaseTimeoutUnit());
+
+            return findMissingKeys(channelName, spokeQueryResults.get(), s3QueryResults.get());
+        } catch (InterruptedException e) {
+            throw new RuntimeInterruptedException(e);
+        } catch (TimeoutException e) {
+            logger.error("s3 verifier timed out while finding missing items, write queue is backing up");
+            statsdReporter.increment(VerifierMetrics.TIMEOUT.getName());
+            return new TreeSet<>();
+        } catch (ExecutionException e) {
+            throw new RuntimeException(e);
+        } finally {
+            Stream.of(spokeQueryResults, s3QueryResults)
+                    .filter(future -> !future.isDone())
+                    .forEach(future -> future.cancel(true));
+        }
+    }
+
+    private SortedSet<ContentKey> findMissingKeys(String channelName, QueryResult spokeQueryResult, QueryResult s3QueryResult) {
+        SortedSet<ContentKey> missingKeys = new TreeSet<>(spokeQueryResult.getContentKeys());
+        missingKeys.removeAll(s3QueryResult.getContentKeys());
+        if (missingKeys.size() > 0) {
+            logger.info("{} missing items found for {}!", missingKeys.size(), channelName);
+            missingKeys.forEach(key -> logger.info("{} missing item {}"));
+        }
+        return missingKeys;
+    }
+
+    private TimeQuery buildTimeQuery(String channelName, MinutePath startPath, MinutePath endPath) {
+        TimeQuery.TimeQueryBuilder builder = TimeQuery.builder()
+                .channelName(channelName)
+                .startTime(startPath.getTime())
+                .unit(TimeUtil.Unit.MINUTES);
+
+        Optional.ofNullable(endPath)
```

Comment:
It is a little opaque what is happening here, or at least it takes a read or two, A comment what the significance of limitKey would be warranted IMHO.